README.md
# axe-runner

[![npm version](https://badge.fury.io/js/axe-runner.svg)](https://badge.fury.io/js/axe-runner)

Puppeteerとaxe-coreで複数ページのアクセシビリティ検証を行い、結果をCSVファイルに出力します。

## 使用方法

事前にテスト対象のURLを[XML形式のサイトマップ](https://developers.google.com/search/docs/advanced/sitemaps/build-sitemap?hl=ja)、もしくは1行1URLでURLを列挙したテキストファイルで準備してください。  
結果は`report`ディレクトリに格納されます。

1. `npm install axe-runner`でインストールします
1. `npx axe-runner [filename of url list]`を実行します（例： `npx axe-runner urls.txt`）

※グローバルにインストールした場合は`npx`をつけずに実行します。

### axeやエミュレートに使用するデバイスの設定

カレントディレクトリの`axe-runner.config.js`に記述します。`-c`オプションで別の場所も指定できます。  
設定方法はWikiの[Config File Example](https://github.com/hideki-a/axe-runner/wiki/Config-File-Example)を参照してください。

`npx axe-runner -c [設定ファイルへの相対パス] [filename of url list]`

### Basic認証への対応

`.env`に認証情報を記述します。

```
BASIC_AUTH_USERNAME=ユーザー名
BASIC_AUTH_PASSWORD=パスワード
```

### コマンドライン初心者向け実行環境

コマンドライン操作が苦手な方のために、ワンクリックでaxe-runnerが実行できる環境を用意しました。Wikiの「[コマンドライン初心者向け実行環境](https://github.com/hideki-a/axe-runner/wiki/%E3%82%B3%E3%83%9E%E3%83%B3%E3%83%89%E3%83%A9%E3%82%A4%E3%83%B3%E5%88%9D%E5%BF%83%E8%80%85%E5%90%91%E3%81%91%E5%AE%9F%E8%A1%8C%E7%92%B0%E5%A2%83)」をご覧ください。

## 補足

- サーバーへのリクエストを行う際、リクエストとリクエストの間に3秒の待機時間を取ります（`interval`で調整可能）

## 参考サイト

開発にあたり参考にしたサイトです。貴重な情報をありがとうございます。

- [axe-core 3.2で検証結果を日本語化する方法 | アクセシビリティBlog | ミツエーリンクス](https://www.mitsue.co.jp/knowledge/blog/a11y/201903/07_1700.html)
- [bjankord/index.js](https://gist.github.com/bjankord/c8afaf345b4499ca3b1267063ce48562)

## 結果を記録したCSVファイル例

![画面キャプチャ：axe-runnerを実行して出力されたCSVの様子](https://user-images.githubusercontent.com/829152/119215222-37302780-bb07-11eb-9ab1-94f40ee7e353.png)


CHANGELOG.md
# axe-runner changelog

## 1.5.8 (2021-05-23)

- 依存モジュールのバージョンを更新

## 1.5.7 (2021-05-22)

- 翻訳言語をlanguageコードで指定できるように変更
- 結果CSVのヘッダー行を日本語化
- 結果CSVの「結果の種別」「影響度」を日本語化

## 1.5.5 (2021-05-01)

- ページ遷移時のオプション（Puppeteerのpage.goto()に渡すオプション）を指定可能にした
- URLファイルの読み込み時・ページ遷移時にエラーが発生した場合の処理を追加

## 1.5.4 (2021-04-28)

- axeを注入する際にエラーが発生するケースへの対応
- 依存モジュールのバージョンを更新（エラーが発生するケースへの対応）

## 1.5.3 (2021-02-25)

- 依存モジュールのバージョンを更新（Message列が空欄になるケースへの対応）

## 1.5.2 (2021-02-25)

- Chromium Helper (Renderer) が多数起動してメモリ消費量が増大する問題を修正

## 1.5.1 (2021-02-18)

- テスト進捗の表示タイミングを調整

## 1.5.0 (2021-02-18)

- リファクタリング（モジュール化により`child_process.spawn()`の利用を廃止）
- テストの進捗表示を追加

### 廃止した機能

- `index.js`を廃止（`npx`による実行と`axe-runner.config.js`によるカスタマイズであらゆるケースに対応できると考えられるため）

## 1.4.0 (2021-02-15)

- エミュレートするデバイスやaxeの設定を記述する`axe-runner.config.js`に対応

## 1.3.0 (2021-02-12)

- 一度に多数のリクエストを発行してしまう問題を修正
- ストリームを用いて結果CSVを出力するように改善
- 結果CSVの文字コードをUTF-8に変更
- エミュレートするデバイスを変更可能にする機能を追加
- HTTP 401, 404のURLを表示する機能を追加

## 1.2.0 (2021-02-11)

- [npmに公開](https://www.npmjs.com/package/axe-runner)
- npmに公開した[@hideki_a/axe-reports](https://www.npmjs.com/package/@hideki_a/axe-reports)を使用するように変更
- Basic認証に対応

## 1.1.0 (2021-02-09)

- `npx axe-runner`でテストが実行できるようにコマンドラインスクリプトを追加
- 1行1URLでURLを列挙したテキストファイルへの対応
- 依存するaxe-core, Puppeteerのバージョンを更新

## 1.0.0 (2019-10-29)

- 基本機能を公開


index.js
require('dotenv').config()
const fs = require('fs');
const deepExtend = require('deep-extend');
const parser = require('fast-xml-parser');
const puppeteer = require('puppeteer');
const AXE_SOURCE = require('axe-core').source;
const AXE_LOCALE_JA = require('axe-core/locales/ja.json');
const AxeReports = require('@hideki_a/axe-reports');

// https://www.npmjs.com/package/sleep
function msleep(n) {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, n);
}

function sleep(n) {
    msleep(n*1000);
}

(async () => {
    const browser = await puppeteer.launch();

    // 設定
    const defaultUserConfigPath = process.cwd() + '/axe-runner.config.js';
    const defaltConfig = {
        config: {
            locale: AXE_LOCALE_JA,
        },
        context: null,
        options: {
            resultTypes: ['violations', 'incomplete', 'inapplicable'],
        },
    };
    let userConfig;
    if (process.argv.length >= 4 && process.argv[3]) {
        const userConfigPath = process.cwd() + '/' + process.argv[3];
        if (fs.existsSync(userConfigPath)) {
            userConfig = require(userConfigPath);
        }
    } else if (fs.existsSync(defaultUserConfigPath)) {
        userConfig = require(defaultUserConfigPath);
    }
    const execConfig = deepExtend(defaltConfig, userConfig);

    // デバイス定義
    let device;
    if (execConfig.device) {
        if (typeof execConfig.device === 'string') {
            device = puppeteer.devices[execConfig.device];
        } else if (typeof execConfig.device === 'object') {
            device = execConfig.device;
        }
    }

    if (!device) {
        const userAgentString = await browser.userAgent();
        device = {
            name: 'Chrome',
            userAgent: userAgentString,
            viewport: {
                width: 1280,
                height: 800,
                deviceScaleFactor: 1,
                isMobile: false,
                hasTouch: false,
                isLandscape: false
            }
        };
    }

    // アクセシビリティテスト開始
    AxeReports.createCsvReportHeaderRow();

    const file = process.argv[2] ? process.argv[2] : 'sitemap.xml';
    const isXML = /\.xml$/.test(file) ? true : false;
    let urls;
    if (isXML) {
        const sitemap = fs.readFileSync(file, { encoding: "utf-8" });
        const sitemapJSON = parser.parse(sitemap);
        urls = sitemapJSON.urlset.url;
    } else {
        urls = fs.readFileSync(file, { encoding: "utf-8" });
        urls = urls.replace(/\r\n?/g, '\n');
        urls = urls.split('\n');
    }

    for (let i = 0, nUrls = urls.length; i < nUrls; i += 1) {
        const url = isXML ? urls[i]['loc'] : urls[i];
        if (!url) {
            continue;
        }

        const page = await browser.newPage();
        await page.setBypassCSP(true);
        page.on('response', response => {
            if (response.url() === url) {
                if (response.status() === 401) {
                    console.error('HTTP 401 Unauthorized -', response.url());
                } else if (response.status() === 404) {
                    console.error('HTTP 404 Not Found -', response.url());
                }
            }
        });

        // リクエスト間で少し間を取る
        if (i !== 0) {
            sleep(3);
        }

        // ページ読み込み
        if (process.env.BASIC_AUTH_USERNAME && process.env.BASIC_AUTH_PASSWORD) {
            await page.authenticate({
                username: process.env.BASIC_AUTH_USERNAME,
                password: process.env.BASIC_AUTH_PASSWORD
            });
        }
        await page.emulate(device);
        await page.goto(`${url}`);

        // axeを注入して実行
        // See also: https://www.mitsue.co.jp/knowledge/blog/a11y/201903/07_1700.html
        await page.evaluate(`eval(${JSON.stringify(AXE_SOURCE)});`);
        const results = await page.evaluate(
            (config, context, options = {}) => {
                const { axe } = window;
                axe.configure(config);
                return axe.run(context || document, options);
            },
            execConfig.config,
            execConfig.context,
            execConfig.options,
        );

        // 結果を確認
        if (results.violations || results.incomplete || results.inapplicable) {
            AxeReports.createCsvReportRow(results);
        }
    }

    await browser.close();
})();


axe-runner.js
const fs = require('fs');
const parser = require('fast-xml-parser');
const puppeteer = require('puppeteer');
const AXE_SOURCE = require('axe-core').source;
const AXE_LOCALE_JA = require('axe-core/locales/ja.json');
const AxeReports = require('axe-reports');

(async () => {
    const promises = [];
    const browser = await puppeteer.launch();
    // const deviceMobile = puppeteer.devices['iPhone 8'];
    const userAgentString = await browser.userAgent();
    const devicePC = {
        'name': 'Chrome',
        'userAgent': userAgentString,
        'viewport': {
            'width': 1280,
            'height': 600,
            'deviceScaleFactor': 1,
            'isMobile': false,
            'hasTouch': false,
            'isLandscape': false
        }
    };

    AxeReports.createCsvReportHeaderRow();

    const sitemap = fs.readFileSync('sitemap.xml', { encoding: "utf-8" });
    const sitemapJSON = parser.parse(sitemap);
    const urls = sitemapJSON.urlset.url;

    for (let i = 0, nUrls = urls.length; i < nUrls; i += 1) {
        const url = urls[i]['loc'];
        promises.push(browser.newPage().then(async page => {
            await page.setBypassCSP(true);

            // ページ読み込み
            await page.emulate(devicePC);
            await page.goto(`${url}`);

            // axeを注入して実行
            await page.evaluate(`eval(${JSON.stringify(AXE_SOURCE)});`);
            const results = await page.evaluate((config, context, options) => {
                const { axe } = window;
                axe.configure(config);
                return axe.run(context || document, options || {});
            }, { locale: AXE_LOCALE_JA }, null, { resultTypes: ['violations', 'incomplete', 'inapplicable'] });

            // 結果を確認
            if (results.violations || results.incomplete || results.inapplicable) {
                AxeReports.createCsvReportRow(results);
            }
        }));
    }

    await Promise.all(promises);
    await browser.close();
})();


lib/runner.js
require('dotenv').config()
const fs = require('fs');
const deepExtend = require('deep-extend');
const parser = require('fast-xml-parser');
const puppeteer = require('puppeteer');
const AXE_SOURCE = require('axe-core').source;
const AXE_LOCALE_JA = require('axe-core/locales/ja.json');
const AxeReports = require('@hideki_a/axe-reports');
const { Console } = require('console');

// https://www.npmjs.com/package/sleep
function msleep(n) {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, n);
}

function sleep(n) {
    msleep(n*1000);
}

function mergeConfig(options) {
    const defaultUserConfigPath = process.cwd() + '/axe-runner.config.js';
    const defaltConfig = {
        config: {
            locale: AXE_LOCALE_JA,
        },
        context: null,
        options: {
            resultTypes: ['violations', 'incomplete', 'inapplicable'],
        },
        interval: 3,
        navigateOptions: {
            waitUntil: 'load',
        },
    };
    let userConfig;
    if (options.config) {
        const userConfigPath = process.cwd() + '/' + options.config;
        if (fs.existsSync(userConfigPath)) {
            userConfig = require(userConfigPath);
        }
    } else if (fs.existsSync(defaultUserConfigPath)) {
        userConfig = require(defaultUserConfigPath);
    }
    const execConfig = deepExtend(defaltConfig, userConfig);
    return execConfig;
}

async function setupDevice(browser, execConfig) {
    let device;
    if (execConfig.device) {
        if (typeof execConfig.device === 'string') {
            device = puppeteer.devices[execConfig.device];
        } else if (typeof execConfig.device === 'object') {
            device = execConfig.device;
        }
    }

    if (!device) {
        const userAgentString = await browser.userAgent();
        device = {
            name: 'Chrome',
            userAgent: userAgentString,
            viewport: {
                width: 1280,
                height: 800,
                deviceScaleFactor: 1,
                isMobile: false,
                hasTouch: false,
                isLandscape: false
            }
        };
    }

    return device;
}

function loadLocale(locale) {
    let json = null;

    if (!/^[a-zA-Z_]{2,4}$/.test(locale)) {
        console.log('An error in the locale specification.');
        return json;
    }

    if (locale !== 'en') {
        try {
            json = require(`axe-core/locales/${locale}.json`);
        } catch {
            console.log('Failed to read the locale file.');
        }
    }

    return json;
}

function readUrls(urlFile, isXML, reject) {
    let urls;
    if (isXML) {
        try {
            const sitemap = fs.readFileSync(urlFile, { encoding: "utf-8" });
            const sitemapJSON = parser.parse(sitemap);
            urls = sitemapJSON.urlset.url;
        } catch {
            console.log('Failed to read the URL file.')
            reject();
        }
    } else {
        try {
            urls = fs.readFileSync(urlFile, { encoding: "utf-8" });
            urls = urls.replace(/\r\n?/g, '\n');
            urls = urls.split('\n');
        } catch {
            console.log('Failed to read the URL file.')
            reject();
        }
    }
    return urls;
}

function runner(urlFile, options, writeStream) {
    return new Promise(async (resolve, reject) => {
        const browser = await puppeteer.launch();
        const execConfig = mergeConfig(options);
        const device = await setupDevice(browser, execConfig);
        let logger;

        if (writeStream) {
            logger = new Console({ stdout: writeStream });
        }

        if (typeof execConfig.config.locale === 'string') {
            execConfig.config.locale = loadLocale(execConfig.config.locale);
        }

        // アクセシビリティテスト開始
        AxeReports.createCsvReportHeaderRow(execConfig.config.locale.lang, logger);

        const isXML = /\.xml$/.test(urlFile) ? true : false;
        const urls = readUrls(urlFile, isXML, reject);
        for (let i = 0, nUrls = urls.length; i < nUrls; i += 1) {
            const url = isXML ? urls[i]['loc'] : urls[i];
            if (!url) {
                continue;
            }

            // リクエスト間で少し間を取ってURLの処理開始
            if (i !== 0) {
                sleep(execConfig.interval);
            }
            console.log(`Accessibility Testing: URL No.${i + 1}`);

            const page = await browser.newPage();
            await page.setBypassCSP(true);
            if (typeof execConfig.device !== 'object') {
                device.viewport.isMobile = false;
            }
            await page.emulate(device);
            page.on('response', response => {
                if (response.url() === url) {
                    if (response.status() === 401) {
                        console.error('HTTP 401 Unauthorized -', response.url());
                    } else if (response.status() === 404) {
                        console.error('HTTP 404 Not Found -', response.url());
                    }
                }
            });

            // ページ読み込み
            if (process.env.BASIC_AUTH_USERNAME && process.env.BASIC_AUTH_PASSWORD) {
                await page.authenticate({
                    username: process.env.BASIC_AUTH_USERNAME,
                    password: process.env.BASIC_AUTH_PASSWORD
                });
            }
            const response = await page.goto(url, execConfig.navigateOptions)
                .catch(() => {
                    console.log('Connection failed. - ' + url);
                });
            if (!response) {
                continue;
            }
            // await page.screenshot({ path: `page_${i + 1}.png`});

            // axeを注入して実行
            // See also: https://www.mitsue.co.jp/knowledge/blog/a11y/201903/07_1700.html
            await page.evaluate(`eval(${JSON.stringify(AXE_SOURCE)});`);
            await page.evaluate(
                (config, context, options = {}) => {
                    const { axe } = window;
                    axe.configure(config);
                    return axe.run(context || document, options);
                },
                execConfig.config,
                execConfig.context,
                execConfig.options,
            ).then(results => {
                // 結果を確認
                if (results && (results.violations || results.incomplete || results.inapplicable)) {
                    AxeReports.createCsvReportRow(results, execConfig.config.locale.lang, logger);
                }
            }).catch(() => {
                console.log('Error occurred while running axe. - ' + url);
            });

            await page.close();
        }

        await browser.close();
        resolve();
    });
}

module.exports = runner;
